File: Stack/evalRPN.java
class Solution {
    private static final Map<String, BiFunction<Integer, Integer, Integer>> opMap = new HashMap<>();
    
    static {
        opMap.put("+", (a, b) -> a+b);
        opMap.put("-",(a, b) -> a-b);
        opMap.put("*",(a, b) -> a*b);
        opMap.put("/",(a, b) -> a/b);
    }
    
    public int evalRPN(String[] tokens) {
        Stack<Integer> st = new Stack<>();
        for(String token: tokens) {
            if(opMap.containsKey(token)) {
                int second = st.peek();
                st.pop();
                int first = st.peek();
                st.pop();
                BiFunction<Integer, Integer, Integer> op = opMap.get(token);
                int res = op.apply(first, second);
                st.push(res);
            } else {
                st.push(Integer.valueOf(token));
            }
        }
        return st.peek();
        
    }
}

File: Stack/evalRPN.cpp
class Solution {
private:   
    int compute(int first, int second, string op) {
        if(op == "+") return first + second;
        if(op == "-") return first - second;
        if(op == "*") return first * second;
        if(op == "/") return first / second;
        
        return 0;
    }
    
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> s;
        set<string> operands= {"+", "-", "*", "/"};
        for(int i = 0; i < tokens.size(); ++i) {
            if(operands.find(tokens[i]) != operands.end() && !s.empty()) {
                int second = s.top();
                s.pop();
                int first = s.top();
                s.pop();
                s.push(compute(first, second, tokens[i]));
            } else {
                s.push(stoi(tokens[i]));
            }
        }
        return s.top();
    }
};
